interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodecompilation/BaseNodeCompiler.scala

```diff
@@ -0,0 +1,325 @@
+        _ ++ _
+      }
+
+    val objCases = caseExpressions.map(_.compiledObject).sequence
+
+    val compilationResult = NodeCompilationResult(
+      expressionTypingInfos,
+      None,
+      expressionCompilation.map(_.validationContext).getOrElse(Valid(ctx)),
+      objExpression.product(objCases),
+      expressionCompilation.flatMap(_.expressionType)
+    )
+
+    combineErrors(compilationResult, additionalValidations.combineAll)
+  }
+
+  def compileFields(
+      fields: List[pl.touk.nussknacker.engine.graph.variable.Field],
+      ctx: ValidationContext,
+      outputVar: Option[OutputVar]
+  )(implicit nodeId: NodeId): NodeCompilationResult[List[variable.Field]] = {
+
+    final case class CompiledRecordField(
+        field: pl.touk.nussknacker.engine.graph.variable.Field,
+        index: Int,
+        typedExpression: TypedExpression
+    )
+
+    val compliedRecordFields = fields.zipWithIndex.map { case (field, index) =>
+      objectParametersExpressionCompiler
+        .compile(field.expression, Some(node.recordValueFieldName(index)), ctx, Unknown)
+        .map(CompiledRecordField(field, index, _))
+    }
+
+    val recordValuesCompilationResult = compliedRecordFields.traverse { validatedField =>
+      validatedField.andThen { compiledField =>
+        Valid(
+          ExpressionCompilerAdapter.ExpressionCompilation(
+            compiledField.field.name,
+            Some(compiledField.typedExpression),
+            Valid(variable.Field(compiledField.field.name, compiledField.typedExpression.expression))
+          )
+        )
+      }
+    }
+
+    val typedFieldsValidations = compliedRecordFields.map { validatedField =>
+      validatedField.map { compiledField =>
+        RecordValidator.IndexedField(
+          compiledField.field.name,
+          compiledField.typedExpression,
+          compiledField.index
+        )
+      }
+    }
+    val additionalValidationResult = RecordValidator.validateRecord(
+      typedFieldsValidations.collect { case Valid(typedField) => typedField }
+    )
+
+    val typedObject = recordValuesCompilationResult
+      .map { fieldsComp =>
+        TypedObjectTypingResult(fieldsComp.map(f => (f.fieldName, f.typingResult)).toMap)
+      }
+      .valueOr(_ => Unknown)
+
+    val fieldsTypingInfo = recordValuesCompilationResult
+      .map { compilations =>
+        compilations.flatMap(_.expressionTypingInfo).toMap
+      }
+      .getOrElse(Map.empty)
+
+    val compiledFields = recordValuesCompilationResult.andThen(_.map(_.validated).sequence)
+
+    val compilationResult = NodeCompilationResult(
+      expressionTypingInfo = fieldsTypingInfo,
+      parameters = None,
+      validationContext = outputVar.map(ctx.withVariable(_, typedObject)).getOrElse(Valid(ctx)),
+      compiledObject = compiledFields,
+      expressionType = Some(typedObject)
+    )
+
+    combineErrors(compilationResult, additionalValidationResult)
+  }
+
+  private object ExpressionCompilerAdapter {
```

Comment:
I don't exactly get what we adapt here? IMO making this `compileExpression` method just 1st level private method would decrease the noise.

Reply:
Yes, it doesn't make much sense besides some code organisation. I wanted to make this code more readable (thats why I extracted this class from the NodeCompiler) and limit the scope at the same time - thats why there's ValidationAdapter, RecordValidator and this ExpressionCompilerAdapter. Do you think other private objects are ok or flattening this structure or doing something else would be better?

Reply:
I see that you already reuse these patterns between record and other base components. IMO we should extract two new files: `RecordValidator` and `(Base)ComponentValidatorHelper`

Reply:
Thanks for the advice, I've extracted these two files. Done